modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+
+    private static final Logger logger = LogManager.getLogger(EnterpriseGeoIpDownloader.class);
+
+    // for overriding in tests
+    static String DEFAULT_MAXMIND_ENDPOINT = System.getProperty(
+        MAXMIND_SETTINGS_PREFIX + "endpoint.default",
+        "https://download.maxmind.com/geoip/databases"
+    );
+    // n.b. a future enhancement might be to allow for a MAXMIND_ENDPOINT_SETTING, but
+    // at the moment this is an unsupported system property for use in tests (only)
+
+    static String downloadUrl(final String name, final String suffix) {
+        String endpointPattern = DEFAULT_MAXMIND_ENDPOINT;
+        if (endpointPattern.contains("%")) {
+            throw new IllegalArgumentException("Invalid endpoint [" + endpointPattern + "]");
+        }
+        if (endpointPattern.endsWith("/") == false) {
+            endpointPattern += "/";
+        }
+        endpointPattern += "%s/download?suffix=%s";
+
+        // at this point the pattern looks like this (in the default case):
+        // https://download.maxmind.com/geoip/databases/%s/download?suffix=%s
+
+        return Strings.format(endpointPattern, name, suffix);
+    }
+
+    static final String DATABASES_INDEX = ".geoip_databases";
+    static final int MAX_CHUNK_SIZE = 1024 * 1024;
+
+    private final Client client;
+    private final HttpClient httpClient;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    // visible for testing
+    protected volatile EnterpriseGeoIpTaskState state;
+    private volatile Scheduler.ScheduledCancellable scheduled;
+    private final Supplier<TimeValue> pollIntervalSupplier;
+    private final Function<String, HttpClient.PasswordAuthenticationHolder> credentialsBuilder;
+
+    EnterpriseGeoIpDownloader(
+        Client client,
+        HttpClient httpClient,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        long id,
+        String type,
+        String action,
+        String description,
+        TaskId parentTask,
+        Map<String, String> headers,
+        Supplier<TimeValue> pollIntervalSupplier,
+        Function<String, HttpClient.PasswordAuthenticationHolder> credentialsBuilder
+    ) {
+        super(id, type, action, description, parentTask, headers);
+        this.client = client;
+        this.httpClient = httpClient;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.pollIntervalSupplier = pollIntervalSupplier;
+        this.credentialsBuilder = credentialsBuilder;
+    }
+
+    void setState(EnterpriseGeoIpTaskState state) {
+        // this is for injecting the state in GeoIpDownloaderTaskExecutor#nodeOperation just after the task instance has been created
+        // by the PersistentTasksNodeService -- since the GeoIpDownloader is newly created, the state will be null, and the passed-in
+        // state cannot be null
+        assert this.state == null;
+        assert state != null;
```

Comment:
Can you add messages to these asserts? Should the second one be an actual check?